### PR TITLE
category_subcategory

### PR DESCRIPTION
**Pull request by Garvshah01**

### CODE REVIEW

File: app/controllers/categories_controller.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class CategoriesController < ApplicationController
+  def index
+    @categories = Category.all.where(super_category_id: nil)
```
Comment (shivamvinsol) [inline app/controllers/categories_controller.rb]: add a scope `root`

File: app/helpers/categories_helper.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+module CategoriesHelper
```
Comment (shivamvinsol) [inline app/helpers/categories_helper.rb]: remove all unused files

File: app/models/category.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Category < ApplicationRecord
+
+  belongs_to :super_category, class_name: "Category", optional: true
```
Comment (shivamvinsol) [inline app/models/category.rb]: rename to `parent`

File: app/models/category.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  has_many :products, dependent: :restrict_with_error
+  has_many :subcategories_products, through: :sub_categories, source: :products, dependent: :restrict_with_error
+  has_many :sub_categories, class_name: "Category", foreign_key: "super_category_id", dependent: :destroy
+
+  validates :name, presence: true
+  validates :name, uniqueness: true, if: :super_category, unless: :name.nil?
```
Comment (shivamvinsol) [inline app/models/category.rb]: `:name.nil?` - does this even work?
use :name?

Comment (shivamvinsol) [reply app/models/category.rb]: allow_nil?

File: app/models/category.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  validates :name, uniqueness: { scope: :super_category_id}, unless: :name.nil?
```
Comment (shivamvinsol) [inline app/models/category.rb]: spacing issues 🤦🏻

File: app/validators/no_child_of_sub_category_validator.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+class NoChildOfSubCategoryValidator < ActiveModel::Validator
+  def validate(record)
```
Comment (shivamvinsol) [inline app/validators/no_child_of_sub_category_validator.rb]: move to model

File: app/models/category.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+  validates_with NoChildOfSubCategoryValidator
```
Comment (shivamvinsol) [inline app/models/category.rb]: no need of separate validator.
cerate custom method

File: app/models/product.rb [hunk 8]
```diff
@@ -25,8 +26,23 @@ class Product < ApplicationRecord
...
   before_validation :set_name_default
   before_validation :set_discount_price
+  after_create :increment_super_category_counter
+  after_destroy :decrement_super_category_counter
 
   private
+
+  def increment_super_category_counter
+    if category.super_category_id
+      category.super_category.increment!(:products_count,1)
```
Comment (shivamvinsol) [inline app/models/product.rb]: spacing

File: config/routes.rb [hunk 9]
```diff
@@ -1,13 +1,13 @@
 Rails.application.routes.draw do
+  get 'categories/index'
```
Comment (shivamvinsol) [inline config/routes.rb]: use RESTFUL routing (resources)

File: db/migrate/20230419082613_create_categories.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+class CreateCategories < ActiveRecord::Migration[7.0]
+  def change
+    create_table :categories do |t|
+      t.string :name
+      t.references :super_category, foreign_key: {to_table: :categories}
```
Comment (shivamvinsol) [inline db/migrate/20230419082613_create_categories.rb:5]: spacing
rename to parent

**Review by shivamvinsol — CHANGES_REQUESTED** [review]

File: app/models/category.rb (hunk 3 above)
Comment (Garvshah01) [reply app/models/category.rb]: done

File: app/models/category.rb (hunk 7 above)
Comment (Garvshah01) [reply app/models/category.rb]: done

File: app/validators/no_child_of_sub_category_validator.rb (hunk 6 above)
Comment (Garvshah01) [reply app/validators/no_child_of_sub_category_validator.rb]: done

File: config/routes.rb (hunk 9 above)
Comment (Garvshah01) [reply config/routes.rb]: done

File: db/migrate/20230419082613_create_categories.rb (hunk 10 above)
Comment (Garvshah01) [reply db/migrate/20230419082613_create_categories.rb:5]: done in new migration

File: app/controllers/categories_controller.rb (hunk 1 above)
Comment (Garvshah01) [reply app/controllers/categories_controller.rb]: done

File: app/models/category.rb (hunk 5 above)
Comment (Garvshah01) [reply app/models/category.rb]: done

File: app/models/product.rb (hunk 8 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/helpers/categories_helper.rb (hunk 2 above)
Comment (Garvshah01) [reply app/helpers/categories_helper.rb]: done

File: Gemfile [hunk 11]
```diff
@@ -77,4 +77,3 @@ gem 'webpacker'
 gem 'bcrypt' , "~> 3.1.7"
```
Comment (shivamvinsol) [inline Gemfile]: change base branch of PR to main

File: app/controllers/categories_controller.rb [hunk 12]
```diff
@@ -0,0 +1,5 @@
+class CategoriesController < ApplicationController
+  def index
+    @categories = Category.all.root
```
Comment (shivamvinsol) [inline app/controllers/categories_controller.rb]: no need of `all`

File: app/models/category.rb [hunk 13]
```diff
@@ -0,0 +1,22 @@
+class Category < ApplicationRecord
+
+  belongs_to :parent_category, class_name: "Category", optional: true
+  has_many :products, dependent: :restrict_with_error
+  has_many :subcategories_products, through: :sub_categories, source: :products, dependent: :restrict_with_error
+  has_many :sub_categories, class_name: "Category", foreign_key: "parent_category_id", dependent: :destroy
```
Comment (shivamvinsol) [inline app/models/category.rb:5]: move this above `subcategories_products`

File: app/models/category.rb [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+  validates :name, presence: true
+  validates :name, uniqueness: true, if: :parent_category, allow_nil: true
+  validates :name, uniqueness: { scope: :parent_category_id }, allow_nil: true
+  validate NoChildOfSubCategoryValidator
```
Comment (shivamvinsol) [inline app/models/category.rb]: whats this??
is this even working?

File: app/models/category.rb [hunk 15]
```diff
@@ -0,0 +1,22 @@
+class Category < ApplicationRecord
+
+  belongs_to :parent_category, class_name: "Category", optional: true
+  has_many :products, dependent: :restrict_with_error
+  has_many :subcategories_products, through: :sub_categories, source: :products, dependent: :restrict_with_error
+  has_many :sub_categories, class_name: "Category", foreign_key: "parent_category_id", dependent: :destroy
+
+  validates :name, presence: true
+  validates :name, uniqueness: true, if: :parent_category, allow_nil: true
+  validates :name, uniqueness: { scope: :parent_category_id }, allow_nil: true
```
Comment (shivamvinsol) [inline app/models/category.rb]: case sensitive?

File: app/models/category.rb [hunk 16]
```diff
@@ -0,0 +1,22 @@
+class Category < ApplicationRecord
+
+  belongs_to :parent_category, class_name: "Category", optional: true
+  has_many :products, dependent: :restrict_with_error
+  has_many :subcategories_products, through: :sub_categories, source: :products, dependent: :restrict_with_error
+  has_many :sub_categories, class_name: "Category", foreign_key: "parent_category_id", dependent: :destroy
+
+  validates :name, presence: true
+  validates :name, uniqueness: true, if: :parent_category, allow_nil: true
```
Comment (shivamvinsol) [inline app/models/category.rb]: move if at last

File: app/models/category.rb [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+  validates :name, uniqueness: { scope: :parent_category_id }, allow_nil: true
+  validate NoChildOfSubCategoryValidator
+
+  scope :root, -> { where(parent_category_id: nil) }
+
+  private
+
+  def ensure_no_child_Of_sub_category
+    if parent_category.parent_category_id
```
Comment (shivamvinsol) [inline app/models/category.rb]: what if parent_category is nil?

File: app/models/category.rb [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+  has_many :sub_categories, class_name: "Category", foreign_key: "parent_category_id", dependent: :destroy
+
+  validates :name, presence: true
+  validates :name, uniqueness: true, if: :parent_category, allow_nil: true
+  validates :name, uniqueness: { scope: :parent_category_id }, allow_nil: true
+  validate NoChildOfSubCategoryValidator
+
+  scope :root, -> { where(parent_category_id: nil) }
+
+  private
```
Comment (shivamvinsol) [inline app/models/category.rb:14]: Should​ ​not​ ​be​ ​able​ ​to​ ​destroy​ ​category​ ​if​ ​any​ ​products​ ​are​ ​associated​ ​with​ ​its ​ ​​ ​​ ​​ ​sub_categories. where the callback for this?

File: app/models/product.rb [hunk 19]
```diff
@@ -25,8 +26,23 @@ class Product < ApplicationRecord
...
   before_validation :set_name_default
   before_validation :set_discount_price
+  after_create :increment_parent_category_counter
+  after_destroy :decrement_parent_category_counter
 
   private
+
+  def increment_parent_category_counter
+    if category.parent_category_id
+      category.parent_category.increment!(:products_count, 1)
```
Comment (shivamvinsol) [inline app/models/product.rb]: use `increment_counter`

File: app/models/product.rb [hunk 20]
```diff
@@ -25,8 +26,23 @@ class Product < ApplicationRecord
...
+    end
+  end
+
+  def decrement_parent_category_counter
+    if category || category.parent_category_id
+      category.parent_category.decrement!(:products_count,1)
```
Comment (shivamvinsol) [inline app/models/product.rb]: `decrement_counter`

File: config/routes.rb [hunk 21]
```diff
@@ -1,13 +1,13 @@
 Rails.application.routes.draw do
+  resources :categories
```
Comment (shivamvinsol) [inline config/routes.rb]: add only those routes we are using

File: test/controllers/categories_controller_test.rb [hunk 22]
```diff
@@ -0,0 +1,8 @@
+require "test_helper"
```
Comment (shivamvinsol) [inline test/controllers/categories_controller_test.rb]: again these files are not deleted?

DELETE all unused files

**Review by shivamvinsol — CHANGES_REQUESTED** [review]

File: app/models/category.rb (hunk 18 above)
Comment (Garvshah01) [reply app/models/category.rb:14]: handled in association  dependent: :restrict_with_error

File: test/controllers/categories_controller_test.rb (hunk 22 above)
Comment (Garvshah01) [reply test/controllers/categories_controller_test.rb]: done

File: config/routes.rb (hunk 21 above)
Comment (Garvshah01) [reply config/routes.rb]: done

File: app/models/product.rb (hunk 20 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 19 above)
Comment (Garvshah01) [reply app/models/product.rb]: done

File: app/models/category.rb (hunk 17 above)
Comment (Garvshah01) [reply app/models/category.rb]: done

File: app/models/category.rb (hunk 16 above)
Comment (Garvshah01) [reply app/models/category.rb]: done

File: app/models/category.rb (hunk 14 above)
Comment (Garvshah01) [reply app/models/category.rb]: name changed

File: app/models/category.rb (hunk 13 above)
Comment (Garvshah01) [reply app/models/category.rb:5]: done

File: app/controllers/categories_controller.rb (hunk 12 above)
Comment (Garvshah01) [reply app/controllers/categories_controller.rb]: done

File: app/models/category.rb (hunk 15 above)
Comment (Garvshah01) [reply app/models/category.rb]: done

File: Gemfile (hunk 11 above)
Comment (Garvshah01) [reply Gemfile]: done

File: app/models/category.rb [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+
+  validates :name, presence: true
+  validates :name, uniqueness: { scope: :parent_category_id, case_sensitive: false }, allow_nil: true
+  validate :ensure_no_child_Of_sub_category
+
+  scope :root, -> { where(parent_category_id: nil) }
+
+  private
+
+  def ensure_no_child_Of_sub_category
```
Comment (shivamvinsol) [inline app/models/category.rb]: why capital O?

File: app/models/product.rb (hunk 19 above)
Comment (shivamvinsol) [reply app/models/product.rb]: where?
do you review your changes?

File: app/models/product.rb (hunk 20 above)
Comment (shivamvinsol) [reply app/models/product.rb]: same issue

File: test/fixtures/categories.yml [hunk 24]
```diff
@@ -0,0 +1,9 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (shivamvinsol) [inline test/fixtures/categories.yml]: still these files are NOT removed

**Review by shivamvinsol — CHANGES_REQUESTED** [review]

File: test/fixtures/categories.yml (hunk 24 above)
Comment (Garvshah01) [reply test/fixtures/categories.yml]: done

File: app/models/category.rb (hunk 23 above)
Comment (Garvshah01) [reply app/models/category.rb]: done

File: test/fixtures/categories.yml (hunk 24 above)
Comment (Garvshah01) [reply test/fixtures/categories.yml:1]: done

**Review by shivamvinsol — APPROVED** [review]

**Pull request merged**